channels/sshagent/client/sshagent_main.c

```diff
@@ -0,0 +1,398 @@
+ *     the RDP client is running.  Each connection by a separate client to
+ *     xrdp-ssh-agent gets a separate DVC invocation.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <pwd.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <winpr/crt.h>
+#include <winpr/synch.h>
+#include <winpr/thread.h>
+#include <winpr/stream.h>
+
+#include "sshagent_main.h"
+#include <freerdp/channels/log.h>
+
+#define TAG CHANNELS_TAG("sshagent.client")
+
+typedef struct _SSHAGENT_LISTENER_CALLBACK SSHAGENT_LISTENER_CALLBACK;
+struct _SSHAGENT_LISTENER_CALLBACK
+{
+	IWTSListenerCallback iface;
+
+	IWTSPlugin* plugin;
+	IWTSVirtualChannelManager* channel_mgr;
+
+        const char *agent_uds_path;
+};
+
+typedef struct _SSHAGENT_CHANNEL_CALLBACK SSHAGENT_CHANNEL_CALLBACK;
+struct _SSHAGENT_CHANNEL_CALLBACK
+{
+	IWTSVirtualChannelCallback iface;
+
+	IWTSPlugin* plugin;
+	IWTSVirtualChannelManager* channel_mgr;
+	IWTSVirtualChannel* channel;
+
+        int agent_fd;
+	HANDLE thread;
+	CRITICAL_SECTION lock;
+};
+
+typedef struct _SSHAGENT_PLUGIN SSHAGENT_PLUGIN;
+struct _SSHAGENT_PLUGIN
+{
+	IWTSPlugin iface;
+
+	SSHAGENT_LISTENER_CALLBACK* listener_callback;
+};
+
+
+/**
+ * Function to open the connection to the sshagent
+ *
+ * @return The fd on success, otherwise -1
+ */
+static int connect_to_sshagent(const char *udspath)
+{
+        int agent_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+        if (agent_fd == -1)
+        {
+                WLog_ERR(TAG, "Can't open Unix domain socket!");
+                return -1;
+        }
+
+        struct sockaddr_un addr;
+        memset(&addr, 0, sizeof(addr));
+        addr.sun_family = AF_UNIX;
+        strncpy(addr.sun_path, udspath, sizeof(addr.sun_path) - 1);
+        int rc = connect(agent_fd, (struct sockaddr*)&addr, sizeof(addr));
+        if (rc != 0)
+        {
+                WLog_ERR(TAG, "Can't connect to Unix domain socket \"%s\"!",
+                         udspath);
+                return -1;
+        }
+
+        return agent_fd;
+}
+
+
+/**
+ * Entry point for thread to read from the ssh-agent socket and forward
+ * the data to RDP
+ *
+ * @return NULL
+ */
+static void *sshagent_read_thread(void *data)
+{
+        SSHAGENT_CHANNEL_CALLBACK *callback = (SSHAGENT_CHANNEL_CALLBACK *)data;
+        BYTE buffer[4096];
+        int going = 1;
+        UINT status = CHANNEL_RC_OK;
+
+        while (going)
+        {
+                int bytes_read = read(callback->agent_fd,
+                                      buffer,
+                                      sizeof(buffer));
+
+                if (bytes_read == 0)
+                {
+                        /* Socket closed cleanly at other end */
+                        going = 0;
+                }
+                else if (bytes_read < 0)
+                {
+                        if (errno != EAGAIN
```

Comment:
I may have missed something, but your socket is opened in a blocking way (and is not changed afterwards) so you will never get a EAGAIN or a EWOULDBLOCK, and the same applies for writes.

Reply:
I've removed them.